File: antson.csharp/antson.csharp/test/io.odysz/jsession/AnSessionReq.cs
using io.odysz.semantic.jprotocol;
using System.Collections.Generic;
using static io.odysz.semantic.jprotocol.AnsonMsg;

namespace io.odysz.semantic.jsession
{
	/// <summary>
	/// <p>Sessin Request<br />
	/// a: see
	/// <see cref="AnSession"/>
	/// </p>
	/// </summary>
	/// <author>odys-z@github.com</author>
	public class AnSessionReq : AnsonBody
	{
		string type;
		public AnSessionReq() : base(null, null)
		{
			type = "io.odysz.semantic.jsession.AnSessionReq";
		}

		/// <summary>Session connection is ignored and controlled by server.</summary>
		/// <param name="parent"/>
		public AnSessionReq(AnsonMsg parent) : base(parent, null)
		{
			type = "io.odysz.semantic.jsession.AnSessionReq";
		}

		internal string uid;

		public string token { get; protected set; }

		// session's DB access is controlled by server
		// internal virtual string token() { return token; }

		public string iv { get; protected set; }

		// internal virtual string iv() { return iv; }

		internal Dictionary<string, object> mds;

		public virtual string md(string k)
		{
			return mds == null ? null : (string)mds[k];
		}

		public virtual AnSessionReq md(string k, string md)
		{
			if (k == null || string.IsNullOrEmpty(md))
			{
				return this;
			}
			if (mds == null)
			{
				mds = new Dictionary<string, object>();
			}
			mds[k] = md;
			return this;
		}

		/// <summary>Format login request message.</summary>
		/// <param name="uid"/>
		/// <param name="tk64"/>
		/// <param name="iv64"/>
		/// <returns>login request message</returns>
		public static AnsonMsg formatLogin(string uid, string tk64, string iv64)
		{
			// AnSessionReq : AnsonBody
			AnsonMsg jmsg = new AnsonMsg(new Port(Port.session), null);
			AnSessionReq itm = new AnSessionReq(jmsg);
			itm.setup(uid, tk64, iv64);
			itm.A("login");

			jmsg.Body((AnsonBody)itm);
			return jmsg;
		}

		private void setup(string logid, string tk64, string iv64)
		{
			uid = logid;
			token = tk64;
			iv = iv64;
		}
	}
}

File: antson.java/src/test/java/io/odysz/semantic/jprotocol/test/AnSessionReq.java
package io.odysz.semantic.jprotocol.test;

import java.util.HashMap;

import io.odysz.common.LangExt;
import io.odysz.semantic.jprotocol.test.Test_AnsonMsg.Port;

/**<p>Sessin Request<br>
 * a = "login" | "logout" | "heartbeat" ...</p>
 * @author odys-z@github.com
 */
public class AnSessionReq extends AnsonBody {
	public AnSessionReq() {
		super(null, null); // session's DB access is controlled by server
	}

	/**Session connection is ignored and controlled by server.
	 * @param parent
	 */
	public AnSessionReq(Test_AnsonMsg<AnSessionReq> parent) {
		super(parent, null); // session's DB access is controlled by server
	}

	public String uid;
	String token;
	public String token() { return token; }
	String iv;
	public String iv() { return iv; }

	HashMap<String,Object> mds;
	public String md(String k) { return mds == null ? null : (String) mds.get(k); }
	public AnSessionReq md(String k, String md) {
		if (k == null || LangExt.isblank(md))
			return this;
		if (mds == null)
			mds = new HashMap<String, Object>();
		mds.put(k, md);
		return this;
	}

	public String uid() { return uid; }

	/**Format login request message.
	 * @param uid
	 * @param tk64
	 * @param iv64
	 * @return login request message
	 */
	public static Test_AnsonMsg<AnSessionReq> formatLogin(String uid, String tk64, String iv64) {
		Test_AnsonMsg<AnSessionReq> jmsg = new Test_AnsonMsg<AnSessionReq>(Port.session);

		AnSessionReq itm = new AnSessionReq(jmsg);
		itm.uid = uid;
		itm.a("login");

		itm.setup(uid, tk64, iv64);

		jmsg.body((AnsonBody)itm);
		return jmsg;
	}

	private void setup(String uid, String tk64, String iv64) {
		this.uid = uid;
		this.token = tk64;
		this.iv = iv64;
	}
}

